### ISSUE TITLE
dotnet core support

### ISSUE THREAD

**Issue by rdghickman** [Recognos/Metrics.NET#32]
Is there currently any plan for the Metrics.NET library to be made compatible with the dotnet core framework? If not, are you open to any PR that may attempt this?

**Comment by niemyjski** [comment]
+1

**Comment by rdghickman** [comment]
Currently I see a bunch of dependencies that are blockers (though these are mostly not for the actual core Metrics code) and a few things in the actual codebase that are not in dotnet core namespaces (Configuration, the MethodImpl.Synchronized attribute, some things in the Environment, etc.). I'm not an expert at porting this stuff but the effort is not completely trivial. There's some question here as to what the project should support going forward (ie. is Mono still relevant if targetting core?). I haven't gotten around to looking at the test code.

**Comment by niemyjski** [comment]
I'd just target only netstandard and be done with it (it should work everywhere).  This site really helps with upgrading packages: http://packagesearch.azurewebsites.net/ More things in environment have been moved around and some will be coming back this fall. If you have any questions, please let me know.

**Comment by rdghickman** [comment]
I tend to agree about only supporting `netstandard`. I'll update this thread if I do end up with a working port of this, but I can't commit to it as of right now.

**Comment by tsibelman** [comment]
I also took a look , it looks the biggest problem is HttpListener that is removed from .net core, it replaced by Kestrel web server.  I think it will be much easier to port Metrics if it could be separated into core project and one or more extensions projects. This way it would be possible to move gradually to .net core, also it will provide leaner and cleaner core library. If maintainers see it as a good way forward I can work on PR with my suggested separation.

**Comment by niemyjski** [comment]
Why would a metrics project have a http listener?

**Comment by tsibelman** [comment]
It used for pulling metrics data from application, also it used by build in web ui

**Comment by tsibelman** [comment]
@PaulParau What do you think about my suggestion would you accept such pull request where I  separate Metrics into core project and one or more extensions projects?

**Comment by PaulParau** [comment]
In principle yes, splitting optional parts from the "core" sounds like a good idea. We had thought about separating components such as the Elastic Search and InfluxDB reporters into separate projects. Not only would this improve the maintainability of the entire project but, like you said, it would make it easier to eventually add dotnet core support.

Do you have an idea about what would have to be separated, i.e. what those extension projects would be? I'm asking because some thing are trivially separable (like the reporters I mentioned) while other parts bring value to the core metrics library and separating them should be carefully considered (e.g. metrics visualization, text reports - which are served by the HttpListener).

I'll also discuss this with the other maintainers and get back with a more concrete response :)

**Comment by PaulParau** [comment]
I've discussed this further with the team: we're OK with removing optional parts (e.g. reporters I mentioned), but we'd rather not remove features like those served by the HttpListener. Without those, the core metrics library would have no built-in reporting support and would need to be used together with another package.

**Comment by tsibelman** [comment]
In this case HttpListener based implementation need to be rewrite to use Kestrel web server for this purpose. I will take a look into how complex it is.

**Comment by mribichich** [comment]
+1

**Comment by alhardy** [comment]
+1

**Comment by alhardy** [comment]
I've copied across the Metrics.NET core project to my aspnet core integration project and stripped it down so it could be upgraded to dotnet core. Stripped down but it works! What's been removed? 

- Remote Metrics related stuff
- HttpReporter (I do still get the /json endpoint using the middleware i've written in aspnet core)
- CSV Reporter
- All the other reporters graphite, influxdb
- Performance counter stuff
- A few other things commented out around properties on the `Environment`

https://github.com/alhardy/aspnet-metrics/commit/9031ae4d821cd3474cb88e1e9b8f6c2031d85190

**Comment by PaulParau** [comment]
While I agree with removing the evidently non-core stuff (reporters - separating them is actually in progress), I'm a bit concerned about the others (e.g. HttpListener, Visualization - I think it's still useful in certain scenarios).

.NET Core support is not a priority for us now (we're focusing right now on getting the next stable release out the door), but we'll look into it and clarify whether we want to go forward with it and the way we want to do it.

**Comment by rudygt** [comment]
I worked on this some time ago, I got the basics working on core using nancyfx + kestrel for the self hosting. the ui works.

there is the branch if someone wants to give it a look

https://github.com/rudygt/Metrics.NET/tree/CoreClrPrototype 

Regards

**Comment by slawomirbrys** [comment]
Is anybody working on porting this lib to .Net Core?

**Comment by niemyjski** [comment]
I'm not sure, I've been trying to gauge when this will be available so I can update our foundatio metrics implementation for this lib (https://github.com/exceptionless/Foundatio#metrics)

**Comment by alhardy** [comment]
@slawomirbrys  I tried by providing a POC. I ended up porting Metrics.NET to https://github.com/alhardy/AppMetrics. Things slowed down since @etishor unfortunately.

My port is a significant refactor, some functionality is removed to support running on core as mentioned in previous comments, however new functionally also exists. 

I've was using Metrics.NET for years and contributed the OWIN package, unfortunately I was not able to get any guidance etc on dotnet core with little feedback when I made efforts to provide a POC.

I also implemented https://www.nuget.org/packages/AspNet.Metrics/ but this limits to the full .net framework.

I think this is an awesome project by @etishor I've found it very useful and am making best attempts in trying to keep it alive.

**Comment by danielmarbach** [comment]
it seems HTTPListener is coming back https://github.com/dotnet/corefx/blob/master/src/System.Net.HttpListener/src/System/Net/HttpListener.cs

https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/dogfooding.md

Maybe targeting .NET Standard 2.0 will be a good option?

**Comment by niemyjski** [comment]
I'd remove it, things like that shouldn't be in the core library, it should
be an extension.

Thanks
-Blake Niemyjski

**Comment by cocowalla** [comment]
@niemyjski I'd have to agree.

Visualisations and an HTTP endpoint are definitely nice to have for *some* systems, but I certainly don't see them as a _core_ part of a metrics library. I struggle to see any reason not to have these as separate packages.

@PaulParau is there likely to be any movement on this in the near future, or is [App.Metrics](https://github.com/alhardy/AppMetrics) now the best option where .NET Standard support is required (which would be most greenfield projects, I would think)?

**Comment by niemyjski** [comment]
@cocowalla you could also check out foundatio :) https://github.com/exceptionless/foundatio#metrics

**Comment by cocowalla** [comment]
@niemyjski actually, I did have a look yesterday :)  But I'm already using Serilog for logging, am using Quartz.net for jobs, have already implemented my own async queue, and have already implemented my own in-memory caching mechanism. The only bit I'm missing is metrics.

Could have saved some time if I'd been aware of foundatio beforehand!

**Comment by alkampfergit** [comment]
Any news for this issue? Actually we really love Metrics.NET library, but without .NET core support we will need to move to another library.

**Comment by niemyjski** [comment]
@gian, We have an implementation for Foundatio for this lib but we support
others as well (https://github.com/FoundatioFx/Foundatio#metrics) perhaps
short term you can move to one of our impls and then move back when it's
supported via the abstraction?

Thanks
-Blake Niemyjski

**Comment by alkampfergit** [comment]
I'll give it a try. :)

Many thanks.

**Comment by alhardy** [comment]
@alkampfergit There is also https://www.app-metrics.io/getting-started/ ;)
